### PR TITLE
ARROW-574: Python: Add support for nested Python lists in Pandas conversion

### PR DESCRIPTION
**Pull request by xhochy**

### CODE REVIEW

File: python/src/pyarrow/adapters/pandas.cc [hunk 1]
```diff
@@ -1648,7 +1649,17 @@ inline Status ArrowSerializer<TYPE>::ConvertTypedLists(
         RETURN_NOT_OK(value_builder->Append(data, size));
       }
     } else if (PyList_Check(objects[i])) {
-      return Status::TypeError("Python lists are not yet supported");
+      PyAcquireGIL lock;
```
Comment (xhochy) [inline python/src/pyarrow/adapters/pandas.cc]: I'm not sure if we should maybe acquire the GIL always in this function instead of only for pylists.

Comment (tebeka) [inline python/src/pyarrow/adapters/pandas.cc]: +1 Note were calling PyList_Check just before we acquire the GIL

File: python/src/pyarrow/adapters/pandas.cc [hunk 2]
```diff
@@ -1648,7 +1649,17 @@ inline Status ArrowSerializer<TYPE>::ConvertTypedLists(
...
+      int64_t size;
+      std::shared_ptr<arrow::DataType> type;
+      RETURN_NOT_OK(list_builder.Append(true));
+      RETURN_NOT_OK(InferArrowType(objects[i], &size, &type));
+      if (type->type != field->type->type) {
```
Comment (tebeka) [inline python/src/pyarrow/adapters/pandas.cc]: Do we want to allow implicit conversion (like int32 to int64)?

File: python/src/pyarrow/adapters/pandas.cc (hunk 1 above)
Comment (xhochy) [inline python/src/pyarrow/adapters/pandas.cc]: Done.

File: python/src/pyarrow/adapters/pandas.cc (hunk 2 above)
Comment (xhochy) [inline python/src/pyarrow/adapters/pandas.cc]: We may allow this in future but currently we are missing a matching converter or alternatively, a convenient cast function for other integer types.

**Review by wesm — APPROVED** [review]
+1. Shoot, I wish I had merged this before doing #375. I will merge this then rebase

**Pull request closed without merging**